language/move-vm/runtime/src/native_functions.rs

```diff
@@ -62,6 +81,40 @@ pub struct NativeContext<'a> {
     data_store: &'a mut dyn DataStore,
     gas_status: &'a GasStatus<'a>,
     resolver: &'a Resolver<'a>,
+    extensions: &'a mut NativeContextExtensions,
+}
+
+/// A data type to represent a heterogeneous collection of extensions which are available to
+/// native functions. A reference to this is passed into the session function execution
+/// entry points.
+#[derive(Default)]
+pub struct NativeContextExtensions {
+    map: HashMap<TypeId, Box<dyn Any>>,
```

Comment:
Partitioning by module seems to be too restrictive. For example, in the async extension, we have a state shared by a set of natives from different modules. The type-based partitioning seems save because the type is fully owned by one given extension, and can be even private to that extension.

Reply:
> can be even private to that extension.
Ahhhhh, makes sense, I like it :)